main/php_syslog.c

```diff
@@ -0,0 +1,74 @@
+/*
+   +----------------------------------------------------------------------+
+   | PHP Version 7                                                        |
+   +----------------------------------------------------------------------+
+   | Copyright (c) 2017 The PHP Group                                |
+   +----------------------------------------------------------------------+
+   | This source file is subject to version 3.01 of the PHP license,      |
+   | that is bundled with this package in the file LICENSE, and is        |
+   | available through the world-wide-web at the following url:           |
+   | http://www.php.net/license/3_01.txt                                  |
+   | If you did not receive a copy of the PHP license and are unable to   |
+   | obtain it through the world-wide-web, please send a note to          |
+   | license@php.net so we can mail you a copy immediately.               |
+   +----------------------------------------------------------------------+
+   | Author: Philip Prindeville <philipp@redfish-solutions.com>           |
+   +----------------------------------------------------------------------+
+*/
+
+/* $Id$ */
+
+#include <stdio.h>
+#include <string.h>
+#include <assert.h>
+#include <stdlib.h>
+#include "php.h"
+#include "php_syslog.h"
+
+#include "zend_smart_string.h"
+
+#ifdef HAVE_STD_SYSLOG
+#define syslog std_syslog
+#endif
+
+PHPAPI void php_syslog(int priority, const char *message) /* {{{ */
+{
+	const char *ptr;
+	unsigned char c;
+	smart_string buf = {0};
+
+	for (ptr = message; ; ++ptr) {
+		c = *ptr;
+		if (c == '\0') {
+			syslog(priority, "%.*s", (int)buf.len, buf.c);
+			break;
+		}
+
+		if (((' ' <= c) && (c <= '~')) || (PG(syslog_highbit) && (c >= 0x80)))
+			smart_string_appendc(&buf, c);
+		else if (c == '\n') {
```

Comment:
At some point one has to draw a line and say "the way this code is being called is broken" and change the calling code, rather than try to accommodate every possible broken caller.

I wrote the original fix to address the issue of the default (built-in) exception handler calling error_log() with a multiline message (i.e. a stack trace).  Ideally, it should have had multiple calls to error_log(), one per line.

Maybe that was my mistake: I should have just fixed the exception handler to break up its messages...